File: java-soa/order-service/src/main/java/com/ecommerce/order/domain/model/OrderStatus.java
package com.ecommerce.order.domain.model;

/**
 * Order Status Enum - Domain Layer
 * 
 * <p><b>Saga Pattern - State Machine</b>:
 * This enum represents the possible states in the order saga.
 * Each transition is triggered by an event from the ESB.</p>
 * 
 * <pre>
 * State Transitions:
 * PENDING → CONFIRMED (when stock.reserved received)
 * PENDING → CANCELLED (when stock.insufficient received)
 * CONFIRMED → SHIPPED (future: when shipping confirms)
 * CONFIRMED → CANCELLED (when user cancels)
 * </pre>
 */
public enum OrderStatus {

    /** Order created, waiting for inventory confirmation */
    PENDING,

    /** Stock reserved, order confirmed */
    CONFIRMED,

    /** Order cancelled due to insufficient stock or user request */
    CANCELLED,

    /** Order shipped to customer */
    SHIPPED,

    /** Order delivered to customer */
    DELIVERED
}

File: nodejs-soa/order-service/src/domain/model/OrderStatus.ts
// =============================================================================
// Order Status - Domain Value Object
// =============================================================================
// Represents the possible states of an order in the Saga lifecycle.
//
// State Machine:
// PENDING → CONFIRMED (when stock is reserved)
// PENDING → CANCELLED (when stock is insufficient)
// CONFIRMED → CANCELLED (when user cancels)
// =============================================================================

export enum OrderStatus {
  /** Order created, waiting for inventory confirmation */
  PENDING = 'PENDING',

  /** Stock reserved successfully, order is confirmed */
  CONFIRMED = 'CONFIRMED',

  /** Order cancelled (insufficient stock or user cancellation) */
  CANCELLED = 'CANCELLED',
}

/**
 * Valid state transitions for the order lifecycle.
 * This enforces the Saga state machine rules.
 */
export const validTransitions: Record<OrderStatus, OrderStatus[]> = {
  [OrderStatus.PENDING]: [OrderStatus.CONFIRMED, OrderStatus.CANCELLED],
  [OrderStatus.CONFIRMED]: [OrderStatus.CANCELLED],
  [OrderStatus.CANCELLED]: [], // Terminal state - no transitions allowed
};

/**
 * Check if a status transition is valid.
 */
export function isValidTransition(from: OrderStatus, to: OrderStatus): boolean {
  return validTransitions[from]?.includes(to) ?? false;
}
